Adds `REAMDE.md` template for all new packages (#1)
pulsar-cooperative/.github: guides/setup-new-package.md

## Configuring the Repository

This repository has several helpful template resources to aide in the initial setup, these resources are available within `./package-template` and show the text that should be added to a package's readme, as well as many of the actions that will be needed.

### Configuring Repository Settings

This setup relates to settings that must be changed on the repository itself, via the GitHub website.

* Add the following values to the repositories secrets:

  - `PPR_AUTH`: This should be a Pulsar Package Registry API key, that allows access to the PPR under this organizations name.
  - `PUBLISHED`: This is a value that represents a boolean (Although it's still just text) that determines if the package has been published to the PPR already. Using this allows us to ensure that automatic version updates aren't done prior to the package being ready, and ensures we don't accidentally try to publish it twice. Prior to the package being published ensure to set it to `false` and once it's been published for the first time set it to `true` (Match the case!).

> * You'll then want to ensure to add `pulsar-package` to the package's 'topic's to help make this package discoverable on GitHub.

Reviewer: ```suggestion
* You'll then want to ensure to add `pulsar-package` to the package's 'topic's on the GitHub UI to help make this package more discoverable.
```

### Configuring Repository Files

From here, you'll want to match the template files as closely as possible.

Copying over the workflows, and adding the text from the readme.

> And important note about these workflow files, is the template contains a `test.yml`, this is used to run automatic tests on the package during every single PR. The package being forked may already have workflow files to do this. So you'll need to investigate every single workflow file already present to determine if they need to be kept, or modified to work on this new org. Some packages may use a different testing schema, that means we can't use the template one, so make sure to take a close look!

Reviewer: ```suggestion
And important note about these workflow files: the template contains a `test.yml` that is used to run automatic tests on the package during every single PR. The package being forked may already have workflow files to do this, so you'll need to investigate every single workflow file already present to determine if they need to be kept, or modified to work on this new org. Some packages may use a different testing scheme, which means we can't use the template one, so make sure to take a close look. It may be wise to look at any other workflow files, if they are present, to make sure everything is accounted for.
```

# Setup a New Package for Pulsar Cooperative

This guide will walk you through the steps needed to configure and setup a brand new forked package on Pulsar Cooperative.

## Renaming the Package/Repository

> Since the original name of the package is reserved on the Pulsar Package Registry, it will need to be renamed, both within the `package.json` as well as the repository name itself (While the name doesn't have to change, it's best practice to keep these names in sync).

Reviewer 2: ```suggestion
Since the original name of the package is reserved on the Pulsar Package Registry, it will need to be renamed, both within the `package.json` as well as the repository name itself (while the name doesn't have to change, it's best practice to keep these names in sync).
```

> When renaming the package the loose convention would be to reuse the original package name, but adding `-pulsar` to the end, such as `[original-name]-pulsar` or `language-cobol-pulsar`.

Reviewer 2: ```suggestion
When renaming the package the loose convention would be to reuse the original package name, but adding `-pulsar` to the end, such as `[original-name]-pulsar` e.g. `language-cobol-pulsar`.
```

## Configuring the Repository

This repository has several helpful template resources to aide in the initial setup, these resources are available within `./package-template` and show the text that should be added to a package's readme, as well as many of the actions that will be needed.

### Configuring Repository Settings

This setup relates to settings that must be changed on the repository itself, via the GitHub website.

> * Add the following values to the repositories secrets/variables:

Reviewer 2: ```suggestion
* Add the following values to the repository's secrets/variables:
```

>   - `PPR_AUTH`: This should be set at the organization level, and does not need to be set per repository. But this should be a Pulsar Package Registry API key, that allows access to the PPR under this organizations name.

Reviewer 2: ```suggestion
  - `PPR_AUTH`: This should be set at the organization level, and does not need to be set per repository. This should be a Pulsar Package Registry API key, that allows access to the PPR under this organizations name.
```

>   - `PUBLISHED`: This is a value that represents a boolean (although it's still just text) that determines if the package has been published to the PPR already. Using this allows us to ensure that automatic version updates aren't done prior to the package being ready, and ensures we don't accidentally try to publish it twice. Prior to the package being published. Don't set this value manually, as it's lack of existence works just as well as creating it manually and setting it to `false`. Otherwise the `publish.yml` GitHub workflow will create this variable as `true` once run.

Reviewer 2: ```suggestion
  - `PUBLISHED`: This is a value that represents a boolean (although it's still just text) that determines if the package has been published to the PPR already. Using this allows us to ensure that automatic version updates aren't done prior to the package being ready, and ensures we don't accidentally try to publish it twice. This value should not be set manually prior to the package being published, as its lack of existence works just as well as creating it manually and setting it to `false`. Otherwise the `publish.yml` GitHub workflow will create this variable as `true` once run.
```

>   - `PUBLISHED`: This is a value that represents a boolean (Although it's still just text) that determines if the package has been published to the PPR already. Using this allows us to ensure that automatic version updates arne't done prior to the package being ready, and ensures we don't accidentally try to publish it twice. Prior to the package being published. Don't set this value manually, as it's lack of existance works just as well as creating it manually and setting it to `false`. Otherwise the `publish.yml` GitHub workflow will create this variable as `true` once run.

Reviewer 2: ```suggestion
  - `PUBLISHED`: This is a value that represents a boolean (although it's still just text) that determines if the package has been published to the PPR already. Using this allows us to ensure that automatic version updates aren't done prior to the package being ready, and ensures we don't accidentally try to publish it twice. This value should not be set manually prior to the package being published, as its lack of existence works just as well as creating it manually and setting it to `false`. Otherwise the `publish.yml` GitHub workflow will create this variable as `true` once run.
```

Reviewer 2: There is something else here I'm not sure about. There is the sentence "Prior to the package being published." and I'm not quite sure where it relates to in this paragraph.

Author: I think this was intended to flow as `Prior to the package being published, don't set this value manually...`

Saying that we shouldn't manually set the variable `PUBLISHED` as it'll be created automatically, and if it's made before the package is every published, it **will** cause the publishing to fail.

Reviewer 2: Gotcha, I've updated my suggestion if this makes sense.

Author: It does make total sense, but since I committed the original suggestion, your edit is seen as outdated. 

If you want feel free to make it again, otherwise in a little bit I can manually add it via a commit

>   - `REPO_NAME`: This value **must** be created manually, and should be the repositories name, which should match the name used on the Pulsar Package Registry.

Reviewer 2: ```suggestion
  - `REPO_NAME`: This value **must** be created manually, and should be the repository's name, which should match the name used on the Pulsar Package Registry.
```

>   - `GITHUB_REPO_TOKEN`: This should be set at the organization level, and does not need to be set per repository. But this should be a GitHub access token with `repo` scope, so that it can modify repository variables as needed.

Reviewer 2: ```suggestion
  - `GITHUB_REPO_TOKEN`: This should be set at the organization level, and does not need to be set per repository. This should be a GitHub access token with `repo` scope, so that it can modify repository variables as needed.
```

> * You'll then want to ensure to add `pulsar-package` to the package's 'topic's on the GitHub UI to help make this package more discoverable.

Reviewer 2: ```suggestion
* You'll then want to ensure to add `pulsar-package` to the package's "topic's" on the GitHub UI to help make this package more discoverable.
```

### Configuring Repository Files

> From here, you'll want to match the template files as closely as possible.
> 
> Copying over the workflows, and adding the text from the readme.

Reviewer 2: ```suggestion
From here, you'll want to match the template files as closely as possible. You will need to copy the workflows and add the text from the readme template to the package readme.
```

> Copying over the workflows, and adding the text from the readme.

Reviewer 2: Not sure what this line relates to?

Author: This is intended to refer to the user that is setting up the repo manually copying over the template workflows, and template text from the readme. That's present within `./package-template/`.

But yeah should probably be a little more clear.

Reviewer 2: Would something like:

> From here, you'll want to match the template files as closely as possible. You will need to copy the workflows and add the text from the readme template to the package readme.

Author: Yeah this makes things much more clear, seems like a needed addition

> And important note about these workflow files: the template contains a `test.yml` that is used to run automatic tests on the package during every single PR. The package being forked may already have workflow files to do this, so you'll need to investigate every single workflow file already present to determine if they need to be kept, or modified to work on this new org. Some packages may use a different testing scheme, which means we can't use the template one, so make sure to take a close look. It may be wise to look at any other workflow files, if they are present, to make sure everything is accounted for.

Reviewer 2: ```suggestion
An important note about these workflow files: the template contains a `test.yml` that is used to run automatic tests on the package during every single PR. The package being forked may already have workflow files to do this, so you'll need to investigate every single workflow file already present to determine if they need to be kept, or modified, to work on this new org. Some packages may use a different testing scheme, which means we can't use the template one, so make sure to take a close look. It may be wise to look at any other workflow files, if they are present, to make sure everything is accounted for.
```
